### ISSUE TITLE
REasonML - Tupple & Record

### ISSUE THREAD

**Issue by phamhoaivu911** [phamhoaivu911/TIL#38]
### 1. Tupple
- Works like array
- Immutable
- Fixed size
- Heterogeneous (items of different types)
We don't access tupple item via index like we do with array. We access it via destructuring.
```reason
let point = (1.0, 1.0);
let (x, y) = point; // x = 1.0, y = 1.0
```
### 2. Record
- Works like object
- Immutable
- fixed in field name and type

We have to define a type for the record before we create a record.

We don't update a record. It's immutable. We create a new object with the updated field.
```reason
type point = {x: float, y: float};
let point = {x: 1.0, y: 1.0}; // Reason will know this is of type point
let another_point = {...point, y: 2.0};
```
#### Puning
Shorcut syntax like ES6, if the value is a variable that has the same name as the field, drop it.
```reason
let x = 1.0;
let y = 1.0;
let point = {x, y}; // same as let point = {x: x, y: y};
```

### Design Decision
In Js, everything are objects. That makes it hard to optimize performance. Sometimes, we only need a simple data structure, not a heavy object. In reason, we have many tools for many different jobs. We have array, list, tupple, record, and object. They have their specific use cases.
Each of them is optimized for the best performance. And Reason forces us to use them the correct way.
Basically, Reason type system enforces the disciplined usage of date structure so that you can guarantee that it will never be accidentally slow.
